### PR TITLE
Graduate component config to stable in kube-scheduler

### PR DESCRIPTION
**Pull request by kerthcet**
Signed-off-by: kerthcet <kerthcet@gmail.com>

#### What type of PR is this?
/kind feature
/sig scheduling 

#### What this PR does / why we need it:
Kube-scheduler Conponent config graduated to Beta since v1.19, it's time to GA the api as planed.

#### Which issue(s) this PR fixes:

Ref: https://github.com/kubernetes/enhancements/issues/785 
https://github.com/kubernetes/enhancements/pull/3315

#### Does this PR introduce a user-facing change?

```release-note
Kube-Scheduler ComponentConfig is graduated to GA, `kubescheduler.config.k8s.io/v1` is available now. 
Plugin `SelectorSpread` is removed in v1.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/785
```

### CODE REVIEW

**Comment by kerthcet** [comment]
Waiting for the PRR review first https://github.com/kubernetes/enhancements/pull/3346#issuecomment-1153758936
/hold

**Comment by kerthcet** [comment]
/hold cancel
As the PRR approved, this pr is read to review. cc @alculquicondor

**Comment by alculquicondor** [comment]
I have some KEPs to review, so I'll leave this for after enhancements freeze.

File: api/api-rules/violation_exceptions.list [hunk 1]
```diff
@@ -343,6 +343,9 @@ API rule violation: list_type_missing,k8s.io/kube-controller-manager/config/v1al
 API rule violation: list_type_missing,k8s.io/kube-controller-manager/config/v1alpha1,PersistentVolumeBinderControllerConfiguration,VolumeHostCIDRDenylist
 API rule violation: list_type_missing,k8s.io/kube-proxy/config/v1alpha1,KubeProxyConfiguration,NodePortAddresses
 API rule violation: list_type_missing,k8s.io/kube-proxy/config/v1alpha1,KubeProxyIPVSConfiguration,ExcludeCIDRs
+API rule violation: list_type_missing,k8s.io/kube-scheduler/config/v1,ExtenderTLSConfig,CAData
+API rule violation: list_type_missing,k8s.io/kube-scheduler/config/v1,ExtenderTLSConfig,CertData
+API rule violation: list_type_missing,k8s.io/kube-scheduler/config/v1,ExtenderTLSConfig,KeyData
```
Comment (alculquicondor) [inline api/api-rules/violation_exceptions.list:348]: These are just `[]byte` fields, which shouldn't require list_type, IMHO. Can you open an issue in kubernetes/kube-openapi to discuss what could/should be done?

File: pkg/scheduler/apis/config/validation/validation.go [hunk 2]
```diff
@@ -134,6 +134,10 @@ var removedPluginsByVersion = []removedPlugins{
 		schemeGroupVersion: v1beta3.SchemeGroupVersion.String(),
 		plugins:            []string{},
 	},
+	{
+		schemeGroupVersion: v1.SchemeGroupVersion.String(),
```
Comment (alculquicondor) [inline pkg/scheduler/apis/config/validation/validation.go:138]: maybe worth removing this logic as we don't plan to remove plugins past v1.

But you could do it as a separate PR (it might make more sense to submit it before this PR).

File: api/api-rules/violation_exceptions.list (hunk 1 above)
Comment (kerthcet) [reply api/api-rules/violation_exceptions.list:348]: Already one there: https://github.com/kubernetes/kube-openapi/issues/281

**Comment by kerthcet** [comment]
From the Graduation Criteria
```
Beta -> GA Graduation
    Deprecation of legacy plugins.
    Minimal changes in the last beta iteration of the API.
```
We still have one plugin `SelectorSpread` to remove, but it will be finished until v1.27 as planned.

@alculquicondor PTAL.

File: api/api-rules/violation_exceptions.list [hunk 3]
```diff
@@ -343,6 +343,9 @@ API rule violation: list_type_missing,k8s.io/kube-controller-manager/config/v1al
 API rule violation: list_type_missing,k8s.io/kube-controller-manager/config/v1alpha1,PersistentVolumeBinderControllerConfiguration,VolumeHostCIDRDenylist
 API rule violation: list_type_missing,k8s.io/kube-proxy/config/v1alpha1,KubeProxyConfiguration,NodePortAddresses
 API rule violation: list_type_missing,k8s.io/kube-proxy/config/v1alpha1,KubeProxyIPVSConfiguration,ExcludeCIDRs
+API rule violation: list_type_missing,k8s.io/kube-scheduler/config/v1,ExtenderTLSConfig,CAData
```
Comment (alculquicondor) [inline api/api-rules/violation_exceptions.list:346]: Or... we could add the `atomic` type.

But that's a nit.

File: pkg/scheduler/apis/config/validation/validation.go (hunk 2 above)
Comment (alculquicondor) [reply pkg/scheduler/apis/config/validation/validation.go]: /hold
for #108444

Otherwise nothing is preventing `SelectorSpread` to be used in v1. Also, we need to update the KEP graduation criteria.

File: cmd/kube-scheduler/app/options/configfile.go [hunk 4]
```diff
@@ -70,8 +71,10 @@ func encodeConfig(cfg *config.KubeSchedulerConfiguration) (*bytes.Buffer, error)
 		encoder = scheme.Codecs.EncoderForVersion(info.Serializer, configv1beta2.SchemeGroupVersion)
 	case configv1beta3.SchemeGroupVersion.String():
 		encoder = scheme.Codecs.EncoderForVersion(info.Serializer, configv1beta3.SchemeGroupVersion)
+	case configv1.SchemeGroupVersion.String():
+		encoder = scheme.Codecs.EncoderForVersion(info.Serializer, configv1.SchemeGroupVersion)
 	default:
-		encoder = scheme.Codecs.EncoderForVersion(info.Serializer, configv1beta3.SchemeGroupVersion)
+		encoder = scheme.Codecs.EncoderForVersion(info.Serializer, configv1.SchemeGroupVersion)
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/options/configfile.go:77]: In https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/785-scheduler-component-config-api#proposal, it says that we would mark v1beta2 as deprecated. It would be useful to emit a log line for that.

And of course, we should update the documentation.

Comment (kerthcet) [reply cmd/kube-scheduler/app/options/configfile.go:77]: Yes, I didn't forget that, I was plan to mark v1beta2 as deprecated in another PR, this PR only focuses on the graduation.

File: pkg/scheduler/apis/config/validation/validation.go (hunk 2 above)
Comment (kerthcet) [reply pkg/scheduler/apis/config/validation/validation.go]: `SelectorSpread` plugin is not default enabled, but can still be reenabled by KubeSchedulerConfiguration, we shouldn't prevent it from being used until we remove it from in-tree plugins totally.

File: api/api-rules/violation_exceptions.list (hunk 3 above)
Comment (kerthcet) [reply api/api-rules/violation_exceptions.list:346]: I'll add the `atomic` listType to config/v1 only.

File: pkg/scheduler/apis/config/validation/validation.go (hunk 2 above)
Comment (alculquicondor) [reply pkg/scheduler/apis/config/validation/validation.go]: We should prevent it from being used. That's what we did from v1beta1 to v1beta2: we disallowed some plugins.

**Comment by kerthcet** [comment]
>  we need to update https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/785-scheduler-component-config-api#proposal to reflect this change.

Created https://github.com/kubernetes/enhancements/pull/3440

File: pkg/scheduler/apis/config/validation/validation_test.go [hunk 5]
```diff
@@ -575,3 +576,284 @@ func TestValidateKubeSchedulerConfigurationV1beta3(t *testing.T) {
...
+	}
+
+	extenderDuplicateBind := validConfig.DeepCopy()
+	extenderDuplicateBind.Extenders[0].BindVerb = "foo"
+	extenderDuplicateBind.Extenders = append(extenderDuplicateBind.Extenders, config.Extender{
+		PrioritizeVerb: "prioritize",
+		BindVerb:       "bar",
+	})
+
+	goodInvalidPlugins := validConfig.DeepCopy()
```
Comment (alculquicondor) [inline pkg/scheduler/apis/config/validation/validation_test.go:750]:
```suggestion
	validPlugins := validConfig.DeepCopy()
```

File: pkg/scheduler/apis/config/validation/validation_test.go [hunk 6]
```diff
@@ -575,3 +576,284 @@ func TestValidateKubeSchedulerConfigurationV1beta3(t *testing.T) {
...
+	goodInvalidPlugins.Profiles[0].Plugins.Score.Enabled = append(goodInvalidPlugins.Profiles[0].Plugins.Score.Enabled, config.Plugin{Name: "PodTopologySpread", Weight: 2})
+
+	badInvalidPlugins := validConfig.DeepCopy()
```
Comment (alculquicondor) [inline pkg/scheduler/apis/config/validation/validation_test.go:753]:
```suggestion
	invalidPlugins := validConfig.DeepCopy()
```

File: pkg/scheduler/apis/config/validation/validation_test.go [hunk 7]
```diff
@@ -575,3 +576,284 @@ func TestValidateKubeSchedulerConfigurationV1beta3(t *testing.T) {
...
+			expectedToFail: true,
+			config:         mismatchQueueSort,
+		},
+		"good-invalid-plugins": {
+			expectedToFail: false,
+			config:         goodInvalidPlugins,
+		},
+		"bad-invalid-plugins": {
+			expectedToFail: true,
+			config:         badInvalidPlugins,
```
Comment (alculquicondor) [inline pkg/scheduler/apis/config/validation/validation_test.go:839]: is there an errorString we should expect?

File: pkg/scheduler/apis/config/validation/validation_test.go (hunk 6 above)
Comment (kerthcet) [reply pkg/scheduler/apis/config/validation/validation_test.go:753]: It's a comparison to `goodInvalidPlugins`.

File: pkg/scheduler/apis/config/validation/validation_test.go (hunk 7 above)
Comment (kerthcet) [reply pkg/scheduler/apis/config/validation/validation_test.go:839]: Add `errorString` validation to all tests.

File: pkg/scheduler/apis/config/validation/validation_test.go [hunk 8]
```diff
@@ -295,7 +315,7 @@ func TestValidateKubeSchedulerConfigurationV1beta2(t *testing.T) {
 			}
 			fmt.Println(errs)
```
Comment (alculquicondor) [inline pkg/scheduler/apis/config/validation/validation_test.go:316]: remove this line, please.

File: pkg/scheduler/apis/config/validation/validation_test.go [hunk 9]
```diff
@@ -569,7 +607,307 @@ func TestValidateKubeSchedulerConfigurationV1beta3(t *testing.T) {
...
+	for name, scenario := range scenarios {
+		t.Run(name, func(t *testing.T) {
+			errs := ValidateKubeSchedulerConfiguration(scenario.config)
+			if errs == nil && scenario.expectedToFail {
+				t.Error("Unexpected success")
+			}
+			if errs != nil && !scenario.expectedToFail {
+				t.Errorf("Unexpected failure: %+v", errs)
+			}
+			fmt.Println(errs)
```
Comment (alculquicondor) [inline pkg/scheduler/apis/config/validation/validation_test.go:908]: remove this line

File: pkg/scheduler/apis/config/validation/validation_test.go [hunk 10]
```diff
@@ -569,7 +607,307 @@ func TestValidateKubeSchedulerConfigurationV1beta3(t *testing.T) {
 			}
 			fmt.Println(errs)
```
Comment (alculquicondor) [inline pkg/scheduler/apis/config/validation/validation_test.go:608]: remove this line.

**Review by alculquicondor — COMMENTED** [review]
just nits now

**Comment by alculquicondor** [comment]
You can already squash

**Comment by kerthcet** [comment]
squashed.

**Comment by kerthcet** [comment]
kindly ping @liggitt as close to the code freeze deadline.

File: staging/src/k8s.io/kube-scheduler/config/v1/types.go [hunk 11]
```diff
@@ -0,0 +1,377 @@
+/*
```
Comment (liggitt) [inline staging/src/k8s.io/kube-scheduler/config/v1/types.go:1]: diff of v1beta3 → v1:

```diff
2c2
< Copyright 2021 The Kubernetes Authors.
---
> Copyright 2022 The Kubernetes Authors.
17c17
< package v1beta3
---
> package v1
366a367
> 	// +listType=atomic
369a371
> 	// +listType=atomic
372a375
> 	// +listType=atomic
```

File: staging/src/k8s.io/kube-scheduler/config/v1/types_pluginargs.go [hunk 12]
```diff
@@ -0,0 +1,225 @@
+/*
```
Comment (liggitt) [inline staging/src/k8s.io/kube-scheduler/config/v1/types_pluginargs.go:1]: no diff from v1beta1 → v1

File: pkg/scheduler/apis/config/validation/validation.go [hunk 13]
```diff
@@ -134,6 +134,10 @@ var invalidPluginsByVersion = []invalidPlugins{
...
+		plugins:            []string{"SelectorSpread"},
```
Comment (liggitt) [inline pkg/scheduler/apis/config/validation/validation.go:139]: is there a test making sure you can't use a SelectorSpread plugin with a v1 config?

File: pkg/scheduler/apis/config/v1/default_plugins.go [hunk 14]
```diff
@@ -0,0 +1,135 @@
+/*
```
Comment (liggitt) [inline pkg/scheduler/apis/config/v1/default_plugins.go:1]: no diff from v1beta3

File: pkg/scheduler/apis/config/v1/defaults.go [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+	if obj.ClientConnection.Burst == 0 {
+		obj.ClientConnection.Burst = 100
+	}
+
+	// Use the default LeaderElectionConfiguration options
+	componentbaseconfigv1alpha1.RecommendedDefaultLeaderElectionConfiguration(&obj.LeaderElection)
+
+	if obj.PodInitialBackoffSeconds == nil {
+		val := int64(1)
+		obj.PodInitialBackoffSeconds = &val
```
Comment (liggitt) [inline pkg/scheduler/apis/config/v1/defaults.go:157]: resync with v1beta3 defaults which picked up use of pointer.Int64

File: pkg/scheduler/apis/config/v1/defaults.go [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+	}
+
+	if obj.PodMaxBackoffSeconds == nil {
+		val := int64(10)
+		obj.PodMaxBackoffSeconds = &val
```
Comment (liggitt) [inline pkg/scheduler/apis/config/v1/defaults.go:162]: resync with v1beta3 defaults which picked up use of pointer.Int64

File: pkg/scheduler/apis/config/v1/defaults.go [hunk 17]
```diff
@@ -0,0 +1,251 @@
...
+	}
+
+	// Enable profiling by default in the scheduler
+	if obj.EnableProfiling == nil {
+		enableProfiling := true
+		obj.EnableProfiling = &enableProfiling
```
Comment (liggitt) [inline pkg/scheduler/apis/config/v1/defaults.go:168]: resync with v1beta3 defaults which picked up use of pointer.BoolPtr

File: pkg/scheduler/apis/config/v1/defaults.go [hunk 18]
```diff
@@ -0,0 +1,251 @@
...
+	}
+
+	// Enable contention profiling by default if profiling is enabled
+	if *obj.EnableProfiling && obj.EnableContentionProfiling == nil {
+		enableContentionProfiling := true
+		obj.EnableContentionProfiling = &enableContentionProfiling
```
Comment (liggitt) [inline pkg/scheduler/apis/config/v1/defaults.go:174]: resync with v1beta3 defaults which picked up use of pointer.BoolPtr

File: pkg/scheduler/apis/config/v1/defaults.go [hunk 19]
```diff
@@ -0,0 +1,251 @@
...
+		}
+	}
+}
+
+func SetDefaults_NodeResourcesBalancedAllocationArgs(obj *configv1.NodeResourcesBalancedAllocationArgs) {
+	if len(obj.Resources) == 0 {
+		obj.Resources = append(obj.Resources,
+			configv1.ResourceSpec{Name: string(v1.ResourceCPU), Weight: 1},
+			configv1.ResourceSpec{Name: string(v1.ResourceMemory), Weight: 1},
+		)
```
Comment (liggitt) [inline pkg/scheduler/apis/config/v1/defaults.go:219]: is there a reason this doesn't use `defaultResourceSpec` like the v1beta3 defaults do?

File: pkg/scheduler/apis/config/v1/defaults.go [hunk 20]
```diff
@@ -0,0 +1,251 @@
+/*
```
Comment (liggitt) [inline pkg/scheduler/apis/config/v1/defaults.go:1]: It looks like there were no changes in effective default values from v1beta3 → v1. Once we release v1, the defaults for it are locked, so were there any TODOs or outstanding issues where a tweak to defaults was desired?

**Review by liggitt — COMMENTED** [review]
API changes lgtm, so marking that as approved

/approve
for API changes

/hold for a few questions about defaults and testing, and a few nits
@alculquicondor can re-review follow-ups and lgtm

File: pkg/scheduler/apis/config/validation/validation.go (hunk 13 above)
Comment (alculquicondor) [reply pkg/scheduler/apis/config/validation/validation.go:139]: yes, it's in pkg/scheduler/apis/config/validation/validation_test.go

File: pkg/scheduler/apis/config/v1/defaults.go (hunk 17 above)
Comment (alculquicondor) [reply pkg/scheduler/apis/config/v1/defaults.go:168]: `pointer.Bool`

File: pkg/scheduler/apis/config/v1/defaults.go (hunk 19 above)
Comment (alculquicondor) [reply pkg/scheduler/apis/config/v1/defaults.go:219]: it looks like it was just updated recently in #110542

File: pkg/scheduler/apis/config/v1/defaults.go [hunk 21]
```diff
@@ -0,0 +1,251 @@
...
+	if obj.MinCandidateNodesPercentage == nil {
+		obj.MinCandidateNodesPercentage = pointer.Int32Ptr(10)
+	}
+	if obj.MinCandidateNodesAbsolute == nil {
+		obj.MinCandidateNodesAbsolute = pointer.Int32Ptr(100)
+	}
+}
+
+func SetDefaults_InterPodAffinityArgs(obj *configv1.InterPodAffinityArgs) {
+	// Note that an object is created manually in cmd/kube-scheduler/app/options/deprecated.go
```
Comment (alculquicondor) [inline pkg/scheduler/apis/config/v1/defaults.go:188]: It looks like this comment is no longer true (the option was removed)

File: pkg/scheduler/apis/config/v1/defaults.go (hunk 20 above)
Comment (alculquicondor) [reply pkg/scheduler/apis/config/v1/defaults.go:1]: The last iteration was v1beta3, introduced in 1.23, which increased some plugin weights https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/785-scheduler-component-config-api#proposal

Are we sure this is enough?

@Huang-Wei @kerthcet @damemi @denkensk for thoughts. Holding the graduation to 1.26 is an option.

File: pkg/scheduler/apis/config/v1/defaults.go (hunk 19 above)
Comment (kerthcet) [reply pkg/scheduler/apis/config/v1/defaults.go:219]: Yes, this pr hanged for a long time.

File: pkg/scheduler/apis/config/v1/defaults.go (hunk 20 above)
Comment (kerthcet) [reply pkg/scheduler/apis/config/v1/defaults.go:1]: My question is if we want to change the plugins and weights(maybe we treat these two things differently) in the future, should we upgrade the CC again, like v1 -> v2alpha1? If so, being cautious is not a bad thing. Maybe we need some feedbacks like a survey.

If no, since I have not gotten any negative feedbacks from the community so far(maybe just missed). From the guides in `Graduation Criteria`,  it seems ready to GA now. 
```
Deprecation of legacy plugins.
Minimal changes in the last beta iteration of the API.
```

Comment (alculquicondor) [reply pkg/scheduler/apis/config/v1/defaults.go:1]: I would prefer not to start a v2 right when we just released v1 :)

I actually expect that most of the world is still running k8s 1.22 or older. Or even if running 1.23, some might still using v1beta2. Since there is no pressure to get this to GA, maybe it's not a bad idea to send a survey to sig-scheduling? The focus of the survey should be on whether user-influenced like pod topology spreading give strong signal compared to internal scores.

Unless people really feel strong about graduating to GA and leaving it up to providers to decide if they want to change the weights for their customers.

Comment (kerthcet) [reply pkg/scheduler/apis/config/v1/defaults.go:1]: It seems we came to the consequence to graduate this to GA.

**Comment by kerthcet** [comment]
Squashed @alculquicondor

**Comment by alculquicondor** [comment]
Thanks @kerthcet

/lgtm

**Comment by kerthcet** [comment]
Still hold now.
